### PR TITLE
Improve search query quality

### PR DESCRIPTION
**Pull request by chethus**
Add some basic filters based on testing queries to improve query quality. In particular require:
- No URLs
- No special characters
- Limited length
- A minimum amount of salience (importance)

Before:
Screenshot 2020-07-27 at 11 50 25 AM
After:
Screenshot 2020-07-27 at 11 45 48 AM

### CODE REVIEW

**Review by nikhilbirmiwal — CHANGES_REQUESTED** [review]
This kinda change seems best accompanied with some test cases.

**Comment by chethus** [comment]
> This kinda change seems best accompanied with some test cases.

Yeah agreed. Quinn is working on tests for analysis today so maybe it would be best for me to wait for that and build on top of it.

**Comment by chethus** [comment]
> > This kinda change seems best accompanied with some test cases.
> 
> Yeah agreed. Quinn is working on tests for analysis today so maybe it would be best for me to wait for that and build on top of it.

Have added some tests!

File: portfolio/src/test/java/FiltersTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import org.junit.Assert;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+public class FiltersTest {
+    @Test
+    public void FilterTest1() throws MalformedURLException{
```
Comment (QuinnDACollins) [inline portfolio/src/test/java/FiltersTest.java]: Might be good to name function exactly what aspect of the filter you're testing? if possible? not sure how specific the filter is so it may not be necessary

**Review by QuinnDACollins — APPROVED** [review]
LGTM, maybe some info on what each test case is trying to accomplish

File: portfolio/src/main/java/com/google/sps/data/SearchQuery.java [hunk 2]
```diff
@@ -28,12 +28,12 @@
     // Make a query based on salience scores given text and a query size.
     public SearchQuery(Video video, int querySize) {
         String queryText;
+        List<String> keywordList;
         if (languageService == null) {
-            queryText = video.getTitle();
+            keywordList = Arrays.asList(video.getTitle().split(" "));
```
Comment (nikhilbirmiwal) [inline portfolio/src/main/java/com/google/sps/data/SearchQuery.java]: Maybe split it on a regular expression of empty space.

File: portfolio/src/main/java/com/google/sps/data/SearchQuery.java [hunk 3]
```diff
@@ -44,15 +44,31 @@ public SearchQuery(Video video, int querySize) {
             AnalyzeEntitiesResponse response = languageService.analyzeEntities(request);
 
             // Take top few distinct words (they are autosorted by decreasing salience).
-            queryText = response.getEntitiesList().stream()
+            keywordList = response.getEntitiesList().stream()
+                // Require a basic amount of salience (importance)
+                .filter(e -> e.getSalience() > 0.12)
```
Comment (nikhilbirmiwal) [inline portfolio/src/main/java/com/google/sps/data/SearchQuery.java:50]: Add a comment how you came up with 0.12.

File: portfolio/src/test/java/FiltersTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        String title = "Taylor Swift - Blank Space";
```
Comment (nikhilbirmiwal) [inline portfolio/src/test/java/FiltersTest.java]: In general, it's better to use clearly mocked titles and descriptions for unit tests.

File: portfolio/src/test/java/FiltersTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        String title = "US surpasses 4 million coronavirus cases";
+        String description = "The surging cases are causing more uncertainty about whether schools will reopen. ABC’s Mona Kosar Abdi reports.";
+        List<String> keywordsList = Arrays.asList((title + description).toLowerCase().split(" "));
+        String query = SearchQuery.filterKeywords(keywordsList, 10);
+        // Filtering again shouldn't change the query.
+        String filteredQuery = refilter(query);
+        assertEquals(query, filteredQuery);
+    }
+
+    public static String refilter(String query) {
```
Comment (nikhilbirmiwal) [inline portfolio/src/test/java/FiltersTest.java]: This was copy/pasted from the code, so unit tests are kinda meaningless. Instead, it would be better to call SearchQuery.filterKeywords.

File: portfolio/src/test/java/FiltersTest.java (hunk 1 above)
Comment (chethus) [reply portfolio/src/test/java/FiltersTest.java]: Yeah it's sort of just testing everything at once.

File: portfolio/src/main/java/com/google/sps/data/SearchQuery.java (hunk 2 above)
Comment (chethus) [reply portfolio/src/main/java/com/google/sps/data/SearchQuery.java:33]: I did it this way so that when I joined with a space after refiltering, the string should be the exact same (if I filter with a regex this will not be the case). Is that fine?

Comment (chethus) [reply portfolio/src/main/java/com/google/sps/data/SearchQuery.java:33]: Nvm I see what you're saying.

File: portfolio/src/test/java/FiltersTest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.junit.Assert;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+public class FiltersTest {
+    @Test
+    public void GeneralFilterTest1() throws MalformedURLException{
```
Comment (nikhilbirmiwal) [inline portfolio/src/test/java/FiltersTest.java]: NIT: Use slightly more descriptive test names. If the only difference is special characters in the mockTitle and mockDescription, then specify that in the test name. Alternatively, merge the two test cases.

File: portfolio/src/test/java/FiltersTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.junit.Assert;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+public class FiltersTest {
```
Comment (nikhilbirmiwal) [inline portfolio/src/test/java/FiltersTest.java:17]: Could we see some tests for the edge cases? Empty titles, empty descriptions, empty keywordLists, etc.

File: portfolio/src/test/java/FiltersTest.java (hunk 6 above)
Comment (chethus) [reply portfolio/src/test/java/FiltersTest.java]: Ok, I think the test cases are quite similar so I'll merge them.

File: portfolio/src/test/java/FiltersTest.java (hunk 7 above)
Comment (chethus) [reply portfolio/src/test/java/FiltersTest.java:17]: Yeah I can do that. Just to provide some context, I discussed with Quinn and we haven't been able to run the Natural Language API with JUnit so I won't be able to test certain parts of the SearchQuery class.

**Review by nikhilbirmiwal — APPROVED** [review]
